### PR TITLE
Initial custom syntax implementation

### PR DESCRIPTION
**Pull request by daytoncf**
Rudimentary implementation for use custom syntax in the spec file. Supports !menu(x) and !ignore keywords.

### CODE REVIEW

File: support.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+// The string will be the line from out.txt that calls the custom syntax, this is so any necessary parameters can be extracted out.
+// The string slice is the slice containing all of the lines output to stdout
+// The integer will be the current position in the above slice that the grader is in
+// Furthermore, the functions will return two integer values.
+// The first integer represents the result score, which will hold either 1 or 0.
+// The second integer will be the new position in the slice containing stdout
+func initSyntaxDictionary() map[string]func(string, []string, int) (string, int, []string) {
+	SyntaxDictionary := make(map[string]func(string, []string, int) (string, int, []string))
+
+	SyntaxDictionary["menu"] = func(menuCall string, StdOutput []string, startPos int) (string, int, []string) {
```
Comment (jacksonwelsh) [inline support.go:18]: would it make more sense to have this function moved into the global namespace? in golang, functions can be stored as variables:

```go
// declare a type for the functions in the array
type SDFunc func(string, []string, int) (string, int, []string)
func simpleSD(s string, ss []string, i int) (string, int, []string) {
  return s, i, ss
}

SD := make(map[string]SDFunc)
SD["simple"] = simpleSD
```

I _think_ this works, not sure though.

Also not blocking.

File: support.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		// Core functionality of ignore.... just skip da line
+
+		return feedback, startPos, StdOutput
+	}
+	return SyntaxDictionary
+}
+
+func extractXValue(s string) int {
+	x_index := strings.Index(s, "(") + 1
+	x_value := int(s[x_index]) - 48 // Ascii subtract
```
Comment (jacksonwelsh) [inline support.go:100]: may be worth changing this to something a little more legible in the future

perhaps use regex to capture between the parens, then use strconv.Atoi to convert the captured string to a number.

Shorter term, we should add input validation. What happens when someone inputs `!menu(z)`?

Comment (jacksonwelsh) [reply support.go:100]: I see error checking is handled outside. Maybe worth adding parameters to this function that can specify the min and max?

File: support.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	return x_value
+}
+
+func hasOptions(output []string, x_value, currPos int) (pos int, pass string) {
```
Comment (jacksonwelsh) [inline support.go:104]: s/hasOptions/menuHasOptions just to make it more clear, unless this is used by other elements?

File: support.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+		if x_value < 1 || x_value > 9 {
+			return "Invalid menu parameter", startPos, StdOutput
+		}
+		// Get current position in stdout
+		curr := startPos
+
+		// Now, evaluate the menu shape.
+		// Check for title
+		if StdOutput[curr] == "" {
+			feedback = "No title" // Check failed
```
Comment (jacksonwelsh) [inline support.go:54]: definitely like how it's generating string feedback as it goes

Comment (jacksonwelsh) [reply support.go:54]: maybe use a slice so we can just append each new piece of feedback?

File: support.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+		currPos++
+	}
+	return currPos, pass
+}
+
+func hasPrompt(Stdout []string, pos int) (result bool, modifiedStdout []string) {
+	modifiedStdout = Stdout
+	result = true
+	if strings.HasSuffix(strings.TrimSpace(Stdout[pos]), ":") || strings.HasSuffix(strings.TrimSpace(Stdout[pos]), "?") {
```
Comment (jacksonwelsh) [inline support.go:134]: helper method opportunity

File: support.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+			modifiedStdout = append(modifiedStdout[:pos+2], modifiedStdout[pos+1:]...)
+			modifiedStdout[pos+1] = misplacedOutput
+		} else if strings.Contains(strings.TrimSpace(Stdout[pos]), "?") {
+			// Split line into two strings, the prompt and the output that was appended to it
+			tempSlice := strings.SplitAfter(Stdout[pos], "?") // temp slice to extract values out of...
+			modifiedStdout[pos] = tempSlice[0]
+			misplacedOutput := strings.TrimSpace(tempSlice[1])
+			modifiedStdout = append(modifiedStdout[:pos+2], modifiedStdout[pos+1:]...)
+			modifiedStdout[pos+1] = misplacedOutput
+		} else {
```
Comment (jacksonwelsh) [inline support.go:155]: repeating logic, helper method opportunity

File: grade.go [hunk 7]
```diff
@@ -72,6 +72,78 @@ func compare(expected, actual string) (bool, string) {
...
+		if i > 0 {
+			position++ // Increment each step after first pass
+		}
+		// if there are no more lines of actual output to compare it to, break
+		if i+1 > len(actualLines) {
+			break
+		}
+
+		// See if line starts with special character indicating use of custom syntax
+		if strings.HasPrefix(line, "!") {
```
Comment (jacksonwelsh) [inline grade.go:131]: also an opportunity for a helper method- something like

```go
package opengrader.directives

type directive int

const (
  MENU directive = iota
  IGNORE directive = iota
  NONE directive = iota
)

func getDirective(line string) (directive)
```

...or something similar. that way we can farm off the logic to the getDirective function and not worry about replicating it elsewhere.

**Review by jacksonwelsh — APPROVED** [review]
Tests look great! :100:

Some other comments, mostly looking towards improving the system as a whole rather than blocking the merge. So just keep those in mind through the next sprint but otherwise :shipit:

**Comment by daytoncf** [comment]
Thanks for the insightful feedback, will work on your comments the next sprint.

File: support.go (hunk 1 above)
Comment (daytoncf) [reply support.go:18]: I agree, this makes more sense. I'll work on an implementation.

**Pull request merged**